kafka-connect-common/src/main/scala/io/lenses/streamreactor/connect/io/text/OptionIteratorAdaptor.scala

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2017-2023 Lenses.io Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.lenses.streamreactor.connect.io.text
+
+class OptionIteratorAdaptor(optional: () => Option[String]) extends Iterator[String] {
+
+  private var lineNumber: Long           = 0L
+  private var nextVal:    Option[String] = Option.empty
+
+  override def hasNext: Boolean = {
+    nextVal = optional()
```

Comment:
I see this is required to initialise nextVal, and avoid initalisation in ctor. But I think it creates a weird behaviour. An iterator hasNext can be called as many times, it's only next:T which mutates. 

A few points on this:
1. Violation of Single Responsibility Principle: The hasNext() method should focus solely on checking whether there is a next element. Reading the next value in this method would introduce additional responsibilities, violating the principle of having a single responsibility for each method.

2. Side Effects and Advancement of the Iterator: Reading the next value within hasNext() would advance the iterator, altering its state before the client calls next(). This can lead to unexpected behavior and make the code harder to reason about.

3.Performance Considerations: Reading the next value in hasNext() can have performance implications, especially if retrieving the next value involves expensive operations. It is generally more efficient to separate the responsibility of advancing the iterator from checking for the next element.


there are a few options.

1. compute the nextVal in the ctor

```
class OptionIteratorAdaptor(optional: () => Option[String]) extends Iterator[String] {

  private var lineNumber: Long           = 0L
  private var nextVal:    Option[String] = optional()

  override def hasNext: Boolean =  nextVal.isDefined
  
  def next(): String = nextVal match {
    case Some(element) =>
      val currentElement = element
      nextVale = optional()
      currentElement
    case None => //raise an error
  }
```

Reply:
Thank you for the advice and the explanation 🙇